warehouse/models/staging/payments/littlepay/stg_littlepay__device_transactions.sql

```diff
@@ -32,12 +32,69 @@ stg_littlepay__device_transactions AS (
         {{ trim_make_empty_string_null('granted_zone_ids') }} AS granted_zone_ids,
         {{ trim_make_empty_string_null('onward_zone_ids') }} AS onward_zone_ids,
-        _line_number,
+        CAST(_line_number AS INTEGER) AS _line_number,
         `instance`,
         extract_filename,
+        {{ extract_littlepay_filename_ts() }} AS littlepay_export_ts,
+        {{ extract_littlepay_filename_date() }} AS littlepay_export_date,
         ts,
-        littlepay_export_ts,
+        -- hash all content not generated by us to enable deduping full dup rows
+        -- hashing at this step will preserve distinction between nulls and empty strings in case that is meaningful upstream
+        {{ dbt_utils.generate_surrogate_key(['participant_id', 'customer_id',
+            'device_transaction_id', 'littlepay_transaction_id', 'device_id', 'device_id_issuer',
+            'type', 'transaction_outcome', 'transction_deny_reason', 'transaction_date_time_utc',
+            'location_id', 'location_scheme', 'location_name', 'zone_id', 'route_id', 'mode',
+            'direction', 'latitude', 'longitude', 'vehicle_id', 'granted_zone_ids',
+            'onward_zone_ids']) }} AS _content_hash,
     FROM source
+),
+
+add_keys_drop_full_dupes AS (
+    SELECT
+        *,
+        -- generate keys now that input columns have been trimmed & cast and files deduped
+        {{ dbt_utils.generate_surrogate_key(['littlepay_export_ts', '_line_number', 'instance']) }} AS _key,
+        device_transaction_id AS _payments_key,
+    FROM clean_columns_and_dedupe_files
+    {{ qualify_dedupe_full_duplicate_lp_rows() }}
+),
+
+stg_littlepay__device_transactions AS (
+    SELECT
+        participant_id, customer_id,
+            device_transaction_id,
+            littlepay_transaction_id,
+            device_id,
+            device_id_issuer,
+            type,
+            transaction_outcome,
+            transction_deny_reason,
+            transaction_date_time_utc,
+            transaction_date_time_pacific,
+            location_id,
+            location_scheme,
+            location_name,
+            zone_id,
+            route_id,
+            mode,
+            direction,
+            latitude,
+            longitude,
+            vehicle_id,
+            granted_zone_ids,
+            onward_zone_ids,
+            _line_number,
+            `instance`,
+            extract_filename,
+            littlepay_export_ts,
+            littlepay_export_date,
+            ts,
+            _key,
+            _payments_key,
+            _content_hash,
+    FROM add_keys_drop_full_dupes
+    -- Some transactions have placeholder information for routes, stops, and directions in their first export,
+    -- then a later export contains the canonical version of the transaction with that information corrected
     QUALIFY ROW_NUMBER() OVER (PARTITION BY littlepay_transaction_id ORDER BY littlepay_export_ts DESC, transaction_date_time_utc DESC) = 1
```

Comment:
In PR description it says that this is one of the tables that still has natural key duplicates, I am particularly interested in that given this extra qualify.... Can you tell why they are happening?

Reply:
This has been superseded by further investigation - no more partial dupes.